util/types/datum.go

```diff
@@ -939,6 +941,66 @@ func (d *Datum) convertToUint(sc *variable.StatementContext, target *FieldType)
 	return ret, nil
 }
 
+func (d *Datum) convertToMysqlTimestamp(sc *variable.StatementContext, target *FieldType) (Datum, error) {
+	var (
+		ret Datum
+		t   Time
+		err error
+	)
+	fsp := DefaultFsp
+	if target.Decimal != UnspecifiedLength {
+		fsp = target.Decimal
+	}
+	loc := sc.TimeZone
+	switch d.k {
+	case KindMysqlTime:
+		t = d.GetMysqlTime()
+		switch {
+		case t.TimeZone == nil:
+			t.TimeZone = loc
+		case t.TimeZone == time.UTC:
+			// Convert to session timezone.
+			if !t.IsZero() {
+				raw, err := t.Time.GoTime(time.UTC)
+				if err != nil {
+					return ret, errors.Trace(err)
+				}
+				converted := raw.In(loc)
+				t.Time = FromGoTime(converted)
```

Comment:
Should we extract line964-969 into a function? This logic we use in many places and if it will update in the future, we can only change one place.